File: .agents/skills/query-sparql/query.py
#!/usr/bin/env python3
"""Send a SPARQL query to the Open Pulse hub gateway (HTTPS).

Reads OPENPULSE_ENDPOINT and OPENPULSE_AUTH (format: user/password;
username ignored, the token is what matters) from .env at the repo root,
or from the environment, and posts to {OPENPULSE_ENDPOINT}/sparql/query.
Set SPARQL_ENDPOINT / SPARQL_AUTH to override the derived values.

Usage:
    python query.py 'SELECT (COUNT(*) AS ?n) WHERE { ?s ?p ?o }'
    python query.py -                          # query from stdin
    python query.py -f path/to/query.rq        # query from file

    # CONSTRUCT/DESCRIBE — switch format to turtle:
    python query.py --accept turtle 'CONSTRUCT { ?s ?p ?o } WHERE { ?s ?p ?o } LIMIT 10'

The default Accept header asks for JSON bindings, which fits SELECT/ASK.
Use --accept {json,csv,xml,turtle} to override. Updates go to /update;
this script only supports /query for safety. Use curl directly if you
need to mutate.
"""
from __future__ import annotations

import argparse
import base64
import json
import os
import socket
import sys
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path

ACCEPT_MAP = {
    "json": "application/sparql-results+json",
    "csv": "text/csv",
    "xml": "application/sparql-results+xml",
    "turtle": "text/turtle",
}


def load_dotenv() -> None:
    for start in (Path.cwd(), Path(__file__).resolve().parent):
        for parent in [start, *start.parents]:
            env = parent / ".env"
            if env.is_file():
                for raw in env.read_text().splitlines():
                    line = raw.strip()
                    if not line or line.startswith("#") or "=" not in line:
                        continue
                    key, _, value = line.partition("=")
                    os.environ.setdefault(key.strip(), value.strip())
                return


def read_query(args: argparse.Namespace) -> str:
    if args.file:
        return Path(args.file).read_text()
    if args.query == "-" or args.query is None:
        return sys.stdin.read()
    return args.query


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("query", nargs="?", help="SPARQL query, or '-' for stdin")
    parser.add_argument("-f", "--file", help="read query from file")
    parser.add_argument("--accept", choices=ACCEPT_MAP, default="json", help="result format (default: json)")
    parser.add_argument("--raw", action="store_true", help="print response body verbatim (default: pretty-print JSON)")
    args = parser.parse_args()

    load_dotenv()
    base = os.environ.get("OPENPULSE_ENDPOINT")
    endpoint = os.environ.get("SPARQL_ENDPOINT") or (base and f"{base.rstrip('/')}/sparql")
    auth = os.environ.get("SPARQL_AUTH") or os.environ.get("OPENPULSE_AUTH")
    if not endpoint or not auth or "/" not in auth:
        print("error: OPENPULSE_ENDPOINT and OPENPULSE_AUTH (user/password) must be set", file=sys.stderr)
        return 2

    user, _, password = auth.partition("/")
    query = read_query(args).strip()
    if not query:
        print("error: empty query", file=sys.stderr)
        return 2

    token = base64.b64encode(f"{user}:{password}".encode()).decode()
    accept = ACCEPT_MAP[args.accept]
    url = f"{endpoint.rstrip('/')}/query"
    body = urllib.parse.urlencode({"query": query}).encode()

    req = urllib.request.Request(
        url,
        data=body,
        headers={
            "Authorization": f"Basic {token}",
            "Content-Type": "application/x-www-form-urlencoded",
            "Accept": accept,
        },
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, timeout=60) as resp:
            raw = resp.read()
    except urllib.error.HTTPError as e:
        print(f"http {e.code}: {e.read().decode(errors='replace')}", file=sys.stderr)
        return 1
    except urllib.error.URLError as e:
        print(f"network error: {e.reason}", file=sys.stderr)
        return 1
    except (TimeoutError, socket.timeout):
        print("network error: request timed out", file=sys.stderr)
        return 1

    if args.raw or args.accept != "json":
        sys.stdout.buffer.write(raw)
        if not raw.endswith(b"\n"):
            sys.stdout.write("\n")
        return 0

    payload = json.loads(raw)
    bindings = payload.get("results", {}).get("bindings")
    if bindings is not None:
        # SELECT — flatten {var: {type, value}} to {var: value}
        rows = [{k: v["value"] for k, v in row.items()} for row in bindings]
        json.dump(rows, sys.stdout, indent=2, default=str)
    else:
        # ASK or unrecognized — print as-is
        json.dump(payload, sys.stdout, indent=2, default=str)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .claude/skills/query-sparql/query.mjs
#!/usr/bin/env node
// Send a SPARQL query to the Open Pulse hub gateway (HTTPS).
//
// Reads OPENPULSE_ENDPOINT and OPENPULSE_AUTH (format: user/password;
// username ignored) from the nearest .env walking up, or from process.env,
// and posts to {OPENPULSE_ENDPOINT}/sparql/query. Set SPARQL_ENDPOINT /
// SPARQL_AUTH to override the derived values.
//
// Usage:
//   node query.mjs 'SELECT (COUNT(*) AS ?n) WHERE { ?s ?p ?o }'
//   node query.mjs -                              # stdin
//   node query.mjs -f path/to/query.rq
//   node query.mjs --accept turtle 'CONSTRUCT ...'
//
// Updates (/update) are deliberately not supported — use curl if you
// need to mutate.

import { readFile, stat } from 'node:fs/promises';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const ACCEPT_MAP = {
	json: 'application/sparql-results+json',
	csv: 'text/csv',
	xml: 'application/sparql-results+xml',
	turtle: 'text/turtle'
};

async function loadDotenv() {
	for (let dir of [process.cwd(), dirname(fileURLToPath(import.meta.url))]) {
		for (let i = 0; i < 10; i++) {
			const envPath = join(dir, '.env');
			try {
				await stat(envPath);
				const text = await readFile(envPath, 'utf8');
				for (const line of text.split('\n')) {
					const trimmed = line.trim();
					if (!trimmed || trimmed.startsWith('#') || !trimmed.includes('=')) continue;
					const idx = trimmed.indexOf('=');
					const key = trimmed.slice(0, idx).trim();
					const value = trimmed.slice(idx + 1).trim();
					if (process.env[key] === undefined) process.env[key] = value;
				}
				return;
			} catch {
				const parent = dirname(dir);
				if (parent === dir) break;
				dir = parent;
			}
		}
	}
}

async function readStdin() {
	let data = '';
	for await (const chunk of process.stdin) data += chunk;
	return data;
}

async function readQuery(argv) {
	const fileFlag = argv.indexOf('-f');
	if (fileFlag !== -1 && argv[fileFlag + 1]) {
		return await readFile(resolve(argv[fileFlag + 1]), 'utf8');
	}
	const positional = argv.find((a, i) => !a.startsWith('-') && argv[i - 1] !== '--accept');
	if (positional === '-' || positional === undefined) return await readStdin();
	return positional;
}

function flag(argv, name) {
	const i = argv.indexOf(name);
	return i === -1 ? null : argv[i + 1];
}

async function main() {
	const argv = process.argv.slice(2);
	await loadDotenv();

	const base = process.env.OPENPULSE_ENDPOINT;
	const endpoint = process.env.SPARQL_ENDPOINT || (base && `${base.replace(/\/$/, '')}/sparql`);
	const auth = process.env.SPARQL_AUTH || process.env.OPENPULSE_AUTH;
	if (!endpoint || !auth || !auth.includes('/')) {
		console.error('error: OPENPULSE_ENDPOINT and OPENPULSE_AUTH (user/password) must be set');
		process.exit(2);
	}

	const [user, ...passwordParts] = auth.split('/');
	const password = passwordParts.join('/');
	const query = (await readQuery(argv)).trim();
	if (!query) {
		console.error('error: empty query');
		process.exit(2);
	}

	const acceptKey = flag(argv, '--accept') || 'json';
	if (!ACCEPT_MAP[acceptKey]) {
		console.error(`error: unknown --accept value '${acceptKey}'`);
		process.exit(2);
	}
	const raw = argv.includes('--raw');

	const token = Buffer.from(`${user}:${password}`).toString('base64');
	const url = `${endpoint.replace(/\/$/, '')}/query`;
	const body = new URLSearchParams({ query }).toString();

	let res;
	try {
		res = await fetch(url, {
			method: 'POST',
			headers: {
				Authorization: `Basic ${token}`,
				'Content-Type': 'application/x-www-form-urlencoded',
				Accept: ACCEPT_MAP[acceptKey]
			},
			body
		});
	} catch (e) {
		console.error(`network error: ${e.message}`);
		process.exit(1);
	}

	if (!res.ok) {
		console.error(`http ${res.status}: ${await res.text()}`);
		process.exit(1);
	}

	const text = await res.text();
	if (raw || acceptKey !== 'json') {
		process.stdout.write(text.endsWith('\n') ? text : text + '\n');
		return;
	}

	const payload = JSON.parse(text);
	const bindings = payload.results?.bindings;
	if (bindings) {
		const rows = bindings.map((row) => Object.fromEntries(Object.entries(row).map(([k, v]) => [k, v.value])));
		process.stdout.write(JSON.stringify(rows, null, 2) + '\n');
	} else {
		process.stdout.write(JSON.stringify(payload, null, 2) + '\n');
	}
}

main();
